Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroClipmapGenerationPass.azsl

```diff
@@ -10,33 +10,59 @@
     RWTexture2DArray<float2> m_macroNormalClipmaps;
 }
 
-[numthreads(32,32,1)]
+[numthreads(THREAD_NUM_X,THREAD_NUM_Y,1)]
 void MainCS(
-    uint3 dispatchThreadID : SV_DispatchThreadID)
+    uint3 groupID : SV_GroupID,
+    uint3 groupThreadID : SV_GroupThreadID)
 {
-    uint2 pixelPosition = dispatchThreadID.xy;
+    uint numberOfGroups = TerrainSrg::m_clipmapData.m_macroDispatchGroupCountX * TerrainSrg::m_clipmapData.m_macroDispatchGroupCountY;
+    uint numberOfGroupThreads = THREAD_NUM_X * THREAD_NUM_Y;
 
     for (uint clipmapLevel = 0; clipmapLevel < TerrainSrg::m_clipmapData.m_macroClipmapStackSize; ++clipmapLevel)
     {
-        uint3 texelIndex = uint3(dispatchThreadID.xy, clipmapLevel);
+        for (uint updateRegionIndex = 0; updateRegionIndex < UpdateRegionMax; ++updateRegionIndex)
+        {
+            uint4 updateRegion = TerrainSrg::m_clipmapData.m_macroClipmapBoundsRegions[clipmapLevel + MacroClipmapStackSizeMax * updateRegionIndex];
+            uint2 size = uint2(updateRegion.z - updateRegion.x, updateRegion.w - updateRegion.y);
+            uint totalNumberOfTexels = size.x * size.y;
+
+            uint totalStride = (totalNumberOfTexels + numberOfGroups - 1) / numberOfGroups;
+            uint totalBegin = min(totalStride * (groupID.x + groupID.y * TerrainSrg::m_clipmapData.m_macroDispatchGroupCountX), totalNumberOfTexels);
+            uint totalEnd = min(totalBegin + totalStride, totalNumberOfTexels);
+
+            uint groupTotal = totalEnd - totalBegin;
+            uint groupStride = (groupTotal + numberOfGroupThreads - 1) / numberOfGroupThreads;
+            uint groupBegin = min(groupStride * (groupThreadID.x + groupThreadID.y * THREAD_NUM_X), groupTotal);
+            uint groupEnd = min(groupBegin + groupStride, groupTotal);
+
+            for (uint i = groupBegin; i < groupEnd; ++i)
+            {
+                uint texelFlatIndex = totalBegin + i;
+                uint2 texelPosition = uint2(texelFlatIndex % size.x + updateRegion.x, texelFlatIndex / size.x + updateRegion.y);
+                uint3 texelIndex = uint3(texelPosition, clipmapLevel);
 
-        float2 worldPosition = GetCurrentWorldPositionFromMacroClipmaps(pixelPosition, clipmapLevel);
-        float clipmapScaleInv = GetMacroClipmapScaleInv(clipmapLevel);
-        float2 positionDdx = ddxPosition(TerrainSrg::m_clipmapData.m_macroClipmapMaxRenderRadius, clipmapScaleInv);
-        float2 positionDdy = ddyPosition(TerrainSrg::m_clipmapData.m_macroClipmapMaxRenderRadius, clipmapScaleInv);
+                float2 worldPosition = GetCurrentWorldPositionFromMacroClipmaps(texelPosition, clipmapLevel);
+                float clipmapScaleInv = GetMacroClipmapScaleInv(clipmapLevel);
+                float2 positionDdx = ddxPosition(TerrainSrg::m_clipmapData.m_macroClipmapMaxRenderRadius, clipmapScaleInv);
+                float2 positionDdy = ddyPosition(TerrainSrg::m_clipmapData.m_macroClipmapMaxRenderRadius, clipmapScaleInv);
```

Comment:
Recomputing the same value here just in X and Y, no? Can save/optimize out the duplicate calculation.